homeassistant/helpers/config_validation.py

```diff
@@ -532,6 +533,72 @@ def url(value: Any) -> str:
     raise vol.Invalid('invalid url')
 
 
+def ip_address(value: Any) -> str:
+    """Validate an IP address (IPv4/IPv6)."""
+    value = str(value)
+    try:
+        ip = ipaddress.ip_address(value)
+    except ValueError:
+        raise vol.Invalid("Invalid IP address: {}.".format(value))
+    return str(ip)
+
+
+def ipv4_address(value: Any) -> str:
+    """Validate an IPv4 address."""
+    value = str(value)
+    try:
+        ip = ipaddress.ip_address(value)
+    except ValueError:
+        raise vol.Invalid("Invalid IPv4 address: {}.".format(value))
+    if ip.version != 4:
+        raise vol.Invalid("IPv{} addresses not allowed.".format(ip.version))
+    return str(ip)
+
+
+def ipv6_address(value: Any) -> str:
+    """Validate an IPv6 address."""
+    value = str(value)
+    try:
+        ip = ipaddress.ip_address(value)
+    except ValueError:
+        raise vol.Invalid("Invalid IPv6 address: {}.".format(value))
+    if ip.version != 6:
+        raise vol.Invalid("IPv{} addresses not allowed.".format(ip.version))
+    return str(ip)
+
+
+def mac48_address(value: Any) -> str:
+    """Validate and normalize a 48-bit MAC address."""
+    mac = str(value).upper()
+    if re.match(r'[\dA-F]{1,2}([-:\s])[\dA-F]{1,2}(\1[\dA-F]{1,2}){4}$', mac):
```

Comment:
Before each of the `if` statements, please add a comment to indicate an example of the pattern being matched.  This will aid with maintainability.

Reply:
Okay.

Reply:
How about that?
```py
def mac48_address(value: Any) -> str:
    """Validate and normalize a 48-bit MAC address."""
    mac = str(value).upper()
    # a1-0b-c3-d4-0e-f6, a1:0b:c3:d4:0e:f6, a1 0b c3 d4 0e f6
    # a1-b-c3-d4-e-f6, a1:b:c3:d4:e:f6, a1 b c3 d4 e f6
    if re.match(r'[\dA-F]{1,2}([-:\s])[\dA-F]{1,2}(\1[\dA-F]{1,2}){4}$', mac):
        octets = [octet.zfill(2) for octet in re.split(r':|-|\s', mac)]
        return '-'.join(octets)
    # a10b-c3d4-0ef6, a10b.c3d4.0ef6, a10b c3d4 0ef6
    if re.match(r'[\dA-F]{4}([-.\s])[\dA-F]{4}\1[\dA-F]{4}$', mac):
        mac = re.sub(r'[-.\s]', '', mac)
        octets = ['{}{}'.format(mac[i], mac[i+1]) for i in range(0, 12, 2)]
        return '-'.join(octets)
    # a10bc3d40ef6
    if re.match(r'[\dA-F]{12}$', mac):
        octets = ["{}{}".format(mac[i], mac[i+1]) for i in range(0, 12, 2)]
        return '-'.join(octets)
    raise vol.Invalid("Invalid MAC address: {}".format(value))
```